resources/lib/api.py

```diff
@@ -178,38 +160,30 @@ def get_episode_id(showid, show_season, show_episode):
             tvshowid=int(showid),
         ))
 
-        if not result.get('result', {}).get('episodes'):
-            return 0
-
         episodeid = 0
-        for episode in result.get('result').get('episodes'):
+        for episode in result.get('result').get('episodes', []):
             if episode.get('episodeid') and episode.get('season') == show_season and episode.get('episode') == show_episode:
                 episodeid = episode.get('episodeid')
 
         return episodeid
 
     def find_next_episode(self, result, current_file, include_watched, current_episode_id):
-        position = 1
-        for episode in result.get('result').get('episodes'):
+        found_match = False
+        episodes = result.get('result', {}).get('episodes', [])
+        for episode in episodes:
             # Find position of current episode
             if current_episode_id == episode.get('episodeid'):
-                # Found a match so get out of here
-                break
-            position += 1
-
-        # Check if it may be a multi-part episode
-        while result.get('result').get('episodes')[position].get('file') == current_file:
-            position += 1
-
-        # Skip already watched episodes?
-        while not include_watched and result.get('result').get('episodes')[position].get('playcount') > 1:
```

Comment:
This looked to me as a bug as `playcount > 1` means it has been watched at least twice, not just once.

Reply:
I'm not sure maybe Kodi defaults to having the playcount as 1

Logically I agree with you but I'm sure we had it in there for a reason or have you confirmed the items in the kodi database that are not watched have a playcount of 0?

Reply:
Playcount really is what it means.
https://forum.kodi.tv/showthread.php?tid=300923

Reply:
Ok cool let's go for it...will merge these changes in now

Reply:
(I just made a small improvement.)
But I would prefer if this gets a bit more testing.

Reply:
No problem I'll hold off for now, soon I'll make a pre-release and upload it to the Emby repository where there are a fair few candidates for testing..prior to releasing to the kodi repo

Reply:
I always felt that Kodi would benefit from a wider testing community so that add-on developers can run test-releases of add-ons one or two weeks before they are released to a wider community. It would have prevent the urllib3 debacle.

A single library update could cause a lot of havoc for users. (Up Next is not at risk though)

Reply:
BTW I still would like to see better integration tests to test fake playback and Up Next behaviour. But it means we have to implement a lot more stub functionality in the Player().

Reply:
Maybe we can have a "TestPlayer" that has all the stubbed functionality and inject that when running the tests...similar ish to the Developer mode. I'm really a Java guy so not sure how/if that dependency can be injected at runtime etc

Reply:
Well, I would like to test them in test/ without requiring Kodi. So we don't even need TestPlayer, just a stub implementation of the test/xbmc.Player is sufficient. I think the difficulty here will be to have this Player run in a thread, then instruct it from the integration tests to start playing, then seek, communicate, and then find out whether Up Next has performed what we expect (based on the configuration set). And for each test we need to reset the Player, and ultimately query its state.

I was a lot more optimistic before I looked at how it worked :-) But I still plan to do this, it's just not happening before making the improvements :-/

Reply:
Hmm, I wanted to test this specific change, and I noticed that this doesn't work. Thinking about this, it is the add-on that proposes the next episode to be played. So this can only work for local files. Am I correct?

In this case I need to take this into account when redesigning the settings, because this is not clear for the user at this time.

I guess we need to make a distinction in the settings between options that affects local playback only , affects some add-ons and affects all add-ons.

Reply:
It would also be nice if the add-on had an icon to access the settings from the add-ons menu.

Reply:
I can confirm that the new implementation works for local playback. I then tried an older implementation, and it failed to work if the playcount was 1. What is strange here is that if the playcount was 1, it did not even propose to play anything. Whereas if the playcount was 2 it would jump over that episode.

So I think we indeed fixed a bug.

Reply:
So maybe we should merge this as this now has become a bugfix PR :-)